### DISCUSSION TITLE
Lifecycle rule - transition to S3 bucket

### DISCUSSION THREAD

**Discussion by madloki** [minio/minio#13630] [Q&A] [answered]
Hello, I have a strange bahaviour when data from my test minio instance are copied trough lifecycle rule into AWS S3 bucket.
On minio, there is a default list of compressed extensisons. After data was succesfully moved to S3 tier, I found this at AWS side:

Instead of test.txt filename, on AWS S3 are just some hashes for directory and filenames.

The filename on S3 is differ, it looks like some kind of compression was applied just on minio side (RLE), so file on AWS S3 is
smaller (and RLE compressed - instead of multiple whitespaces are some less control characters)

Minio:

00000000  74 68 69 73 20 69 73 20  61 20 74 65 73 74 0a 0a  |this is a test..|
00000010  61 6e 6f 74 68 65 72 20  6c 69 6e 65 20 61 64 64  |another line add|
00000020  65 64 0a 0a 79 65 74 20  61 6e 6f 74 68 65 72 20  |ed..yet another |
00000030  6c 69 6e 65 0a 0a 76 65  72 73 69 6f 6e 20 34 0a  |line..version 4.|
00000040  0a 76 65 72 73 69 6f 6e  20 35 0a 0a 76 65 72 73  |.version 5..vers|
00000050  69 6f 6e 20 36 0a 0a 76  65 72 73 69 6f 6e 20 37  |ion 6..version 7|
00000060  0a 0a 76 65 72 73 69 6f  6e 20 38 0a 0a 76 65 72  |..version 8..ver|
00000070  73 69 6f 6e 20 39 0a 0a  76 65 72 73 69 6f 6e 20  |sion 9..version |
00000080  31 30 0a 0a 76 65 72 73  69 6f 6e 20 31 31 20 61  |10..version 11 a|
00000090  6e 64 20 31 32 20 2b 20  31 33 20 2b 20 31 34 20  |nd 12 + 13 + 14 |
000000a0  2b 20 31 35 20 2b 20 31  36 0a                    |+ 15 + 16.|
000000aa

AWS S3:

00000000  ff 06 00 00 53 32 73 54  77 4f 00 7a 00 00 5c f4  |....S2sTwO.z..\.|
00000010  89 f2 aa 01 9c 74 68 69  73 20 69 73 20 61 20 74  |.....this is a t|
00000020  65 73 74 0a 0a 61 6e 6f  74 68 65 72 20 6c 69 6e  |est..another lin|
00000030  65 20 61 64 64 65 64 0a  0a 79 65 74 20 2e 18 00  |e added..yet ...|
00000040  28 0a 0a 76 65 72 73 69  6f 6e 20 34 19 0b 00 35  |(..version 4...5|
00000050  15 00 02 00 36 15 00 02  00 37 15 00 02 00 38 15  |....6....7....8.|
00000060  00 02 00 39 15 00 02 04  31 30 1d 0c 30 31 20 61  |...9....10..01 a|
00000070  6e 64 20 31 32 20 2b 20  31 33 01 05 00 34 01 00  |nd 12 + 13...4..|
00000080  18 35 20 2b 20 31 36 0a                           |.5 + 16.|
00000088

**Comment by harshavardhana** [comment, marked as answer]
If data is transitioned from MinIO then it will have a custom format.

**Comment by madloki** [comment]
Hello,
           many thanks for Your attention and sorry for my english.


If data are not the same, because aws s3 don't recognize that minio format
or AWS is not able to handle it properly, what is purpose of s3 tier
support then ?

I suppose that on-fly decompression to the tier defined as AWS S3 should
fix that easily ?

Did simple disabling predefined extensions and mime-type (disabling
compression) solve that situation ? That should be OK for me, if mentioned
somewhere.

Same story, if s3 tuer us used, mc don't support remote (aws) storage
classes (just 2 minio storage native classes). Reason  ? The tier is
defined like AWS S3 properly, so it is on aws s3 to check that (remote)
storage class validity.

Leave the AWS S3 support now.

I noticed one another file corruption detected using default compression on
plaintext file, just using copy a file trough mc to minio and get it back
using mc or web ui  from the same minio server. In that case on the end of
file was a lot of null characters appended, so wrong size and crc appears
on my workstation.

It was just one file within ~30000 of other text files handled the same way
to and from minio. Anyway, is it possible that RLE handling has some bug
inside ?

Dne st 10. 11. 2021 17:17 uživatel Harshavardhana ***@***.***>
napsal:

**Reply by harshavardhana** [reply]
You cannot use AWS directly with MinIO transition @madloki you need to go through MinIO.

**Comment by madloki** [comment]
Hello, i'm a bit confused now. So transitions are not allowed to minio
configured AWS S3 tier at all ?

In the UI in tiers section is written (cite):

"Learn more about TIERS
Tiers are used by the MinIO Object Lifecycle Management which allows
creating rules for time or date based automatic transition or expiry of
objects. For object transition, MinIO automatically moves the object to a
configured remote storage tier."

When creating a new one tier, I must choose between this types: "minio",
"AWS S3", "GCS", "Azure"

So it is just a fake statement, or future planed feature and improper
documentation also ?
See this:
https://docs.min.io/minio/baremetal/lifecycle-management/transition-objects-to-s3.html#minio-lifecycle-management-transition-to-s3



Dne st 10. 11. 2021 18:33 uživatel Harshavardhana ***@***.***>
napsal:

**Reply by klauspost** [reply]
@madloki You are welcome to use S3 as a tier, however you should not access data directly on S3, only through MinIO, which contains the metadata for it.

What you are seeing is the compressed file, as stored on the MinIO backend. I believe using encryption would have the same effect.

**Comment by madloki** [comment]
Hello, many thanks, maybe I'm closer to understand now. So data transited to AWS S3 tier should be still accessible and visible trough minio in their original bucket, they are just physically stored (except of metadata) on S3 AWS ? In that case, naming & format in which the data are located in AWS S3 as they are accessible trough AWS S3 make no sense (and there is no need for accessing them trough S3 API/UI). Big sorry to harshavardhana then, was my fault (in understanding of the principle)

So if I have an scenario, when I want to have local fast minio1 cluster as direct target for apps producing data, then other minio2 cluster for processing (reducing and counting on them), and at the end I want to move processed data amount (and the new born results) from minio2 to the S3 bucket with  AWS specific storage class (to pay less for them), how to use minio (maybe mc or API is also necessary ?) in that use case  ?

Mean 2 days old and marked-A data delete from minio1, marked-B data older than 2 days move from minio1 to minio2 (delete them from minio1), and 2 days old  marked-B and marked-C data tier from minio2 to S3 bucket, leaving just metadata of them on the minio2 storage ?

**Reply by klauspost** [reply]
@madloki Correct, Tiering is meant to be transparent from the S3 API side, so you don't need to worry if and when an object has been transitioned.

If you want to move your data to another cluster after processing I propose you do that when processing has finished. So implement that functionality in your processing software.

**Comment by madloki** [comment]
Hello,
          OK, thank both of You very much, and sorry for my lazy reading of the documentation.

**Reply by ravindk89** [reply]
We've made some adjustments to the docs since you last read them that might also further clarify the behavior you experienced. We're working on rapid iteration and improvements, so your feedback is actually helpful, and not lazy at all :)
